src/ggrc/login/appengine.py

```
@@ -14,30 +14,64 @@
 E.g., ``login: required`` must be specified *at least* for the '/login' route.
 """
 
+import json
+
 from google.appengine.api import users
-from ggrc.login.common import find_or_create_user_by_email, get_next_url
+import flask
 import flask_login
-from flask import url_for, redirect, request, session, flash
+
+from ggrc.login import common
+from ggrc.models import all_models
+from ggrc import settings
 
 
 def get_user():
+  """Get current user using appengine authentication."""
   ae_user = users.get_current_user()
   email = ae_user.email()
   nickname = ae_user.nickname()
-  user = find_or_create_user_by_email(email, name=nickname)
+  user = common.find_or_create_user_by_email(email, name=nickname)
   return user
 
+
 def login():
+  """Log in current user."""
   user = get_user()
   if user.system_wide_role != 'No Access':
     flask_login.login_user(user)
-    return redirect(get_next_url(request, default_url=url_for('dashboard')))
-  else:
-    flash(u'You do not have access. Please contact your administrator.', 'alert alert-info')
-    return redirect('/')
+    return flask.redirect(common.get_next_url(
+        flask.request, default_url=flask.url_for('dashboard')))
+
+  flask.flash(u'You do not have access. Please contact your administrator.',
+              'alert alert-info')
+  return flask.redirect('/')
+
 
 def logout():
   flask_login.logout_user()
-  return redirect(
-    users.create_logout_url(
-      get_next_url(request, default_url=url_for('index'))))
+  return flask.redirect(users.create_logout_url(common.get_next_url(
+      flask.request, default_url=flask.url_for('index'))))
+
+
+def request_loader(request):
+  """Get the user provided in X-GGRC-user if whitelisted Appid provided."""
+
+  whitelist = settings.ALLOWED_QUERYAPI_APP_IDS
+  inbound_appid = request.headers.get("X-Appengine-Inbound-Appid")
+  if inbound_appid not in whitelist:
+    # don't check X-GGRC-user if the request comes from an untrusted source
+    return None
```

I don't think we should silently ignore if `X-Appengine-Inbound-Appid` is set. If the app is not in the whitelist we should raise a BAD REQUEST explaining to the user that `X-Appengine-Inbound-Appid` is not in the `ALLOWED_QUERYAPI_APP_IDS`. This will help developers A LOT when debugging issues with their setup (it would even help me testing out this PR).